pkg/client/localdocker.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	apiv1 "github.com/google/cloud-android-orchestration/api/v1"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	hoclient "github.com/google/android-cuttlefish/frontend/src/libhoclient"
+	"github.com/hashicorp/go-multierror"
+)
+
+const LocalDockerServiceURL = "local://docker"
+
+type localDockerClient struct {
+	*ClientOptions
+	dcli client.Client
+}
+
+func NewDockerClient(opts *ClientOptions) (Client, error) {
+	dcli, err := client.NewClientWithOpts(client.FromEnv)
+	if err != nil {
+		return nil, err
+	}
+	return &localDockerClient{ClientOptions: opts, dcli: *dcli}, nil
+}
+
+const (
+	defaultLocalDockerImage = "cuttlefish-orchestration:latest"
+	createdByCVDRLabel      = "created_by_cvdr"
+	hoSrvPort               = "2080/tcp"
+)
+
+func (s *localDockerClient) CreateHost(req *apiv1.CreateHostRequest) (*apiv1.HostInstance, error) {
+	ctx := context.TODO()
+	config := &container.Config{
+		Image: defaultLocalDockerImage,
+		Labels: map[string]string{
+			createdByCVDRLabel: "",
+		},
+	}
+	hostConfig := &container.HostConfig{
+		Privileged:   true,
+		PortBindings: nat.PortMap{hoSrvPort: []nat.PortBinding{{HostPort: ""}}},
+	}
+	res, err := s.dcli.ContainerCreate(ctx, config, hostConfig, nil, nil, "")
+	if err != nil {
+		return nil, fmt.Errorf("container create failed: %w", err)
+	}
+	err = s.dcli.ContainerStart(ctx, res.ID, types.ContainerStartOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("container start failed: %w", err)
+	}
```

Comment:
I think lots of code in this file is duplicated to `pkg/app/instances/docker.go`, and I think it's better to make library which both `pkg/app/instances/docker.go` and `pkg/client/localdocker.go` can use. I'm little afraid of the execution flow like container config between remote docker and local docker being distinguished even for unexpected ones in the future.

Reply:
Or, is there any reason for not making such library?

Reply:
there's some duplicated code, however, let's not constrain ourselves by having a rushed refactoring at this point and have the ability of using the docker library independently as there are some key difference in its use, for example: host port mapping and the using/not using the Operations object.

Reply:
no longer relevant.